onvm/onvm_mgr/main.c

```diff
@@ -373,57 +404,96 @@ main(int argc, char *argv[]) {
         signal(SIGINT, handle_signal);
         signal(SIGTERM, handle_signal);
 
+        struct queue_mgr *tx_mgr[tx_lcores];
+        struct queue_mgr *rx_mgr[rx_lcores];
+        struct wakeup_thread_context *wakeup_ctx[ONVM_NUM_WAKEUP_THREADS];
+
         for (i = 0; i < tx_lcores; i++) {
-                struct queue_mgr *tx_mgr = calloc(1, sizeof(struct queue_mgr));
-                tx_mgr->mgr_type_t = MGR;
-                tx_mgr->id = i;
-                tx_mgr->tx_thread_info = calloc(1, sizeof(struct tx_thread_info));
-                tx_mgr->tx_thread_info->port_tx_bufs = calloc(RTE_MAX_ETHPORTS, sizeof(struct packet_buf));
-                tx_mgr->nf_rx_bufs = calloc(MAX_NFS, sizeof(struct packet_buf));
-                tx_mgr->tx_thread_info->first_nf = RTE_MIN(i * nfs_per_tx + 1, (unsigned)MAX_NFS);
-                tx_mgr->tx_thread_info->last_nf = RTE_MIN((i + 1) * nfs_per_tx + 1, (unsigned)MAX_NFS);
+                tx_mgr[i] = rte_calloc(NULL, 1, sizeof(struct queue_mgr), RTE_CACHE_LINE_SIZE);
+                if (tx_mgr[i] == NULL) {
+                        RTE_LOG(ERR, APP, "Can't allocate queue_mgr struct\n");
+                        onvm_main_free(tx_lcores, rx_lcores, tx_mgr, rx_mgr, wakeup_ctx);
+                        return -1;
+                }
+                tx_mgr[i]->mgr_type_t = MGR;
+                tx_mgr[i]->id = i;
+                tx_mgr[i]->tx_thread_info = rte_calloc(NULL, 1, sizeof(struct tx_thread_info), RTE_CACHE_LINE_SIZE);
+                if (tx_mgr[i]->tx_thread_info == NULL) {
+                        RTE_LOG(ERR, APP, "Can't allocate tx_thread_info struct\n");
+                        onvm_main_free(tx_lcores,rx_lcores, tx_mgr, rx_mgr, wakeup_ctx);
+                        return -1;
+                }
+                tx_mgr[i]->tx_thread_info->port_tx_bufs =
+                    rte_calloc(NULL, RTE_MAX_ETHPORTS, sizeof(struct packet_buf), RTE_CACHE_LINE_SIZE);
+                if (tx_mgr[i]->tx_thread_info->port_tx_bufs == NULL) {
+                        RTE_LOG(ERR, APP, "Can't allocate packet_buf struct\n");
+                        onvm_main_free(tx_lcores,rx_lcores, tx_mgr, rx_mgr, wakeup_ctx);
+                        return -1;
+                }
+                tx_mgr[i]->nf_rx_bufs = rte_calloc(NULL, MAX_NFS, sizeof(struct packet_buf), RTE_CACHE_LINE_SIZE);
+                if (tx_mgr[i]->nf_rx_bufs == NULL) {
+                        RTE_LOG(ERR, APP, "Can't allocate packet_buf struct\n");
+                        onvm_main_free(tx_lcores,rx_lcores, tx_mgr, rx_mgr, wakeup_ctx);
```

Comment:
I think that calling onvm_main_free after every check may be redundant, could we maybe set a flag and move that call out of the loop so that it's called only once? We should fail on any unsuccessful allocation anyway.

Reply:
I think calling onvm_main_free followed by RTE_EXIT would be best? If not I could also see using a flag or a goto label as another option.

Reply:
@bdevierno1 you probably want to just set a flag and break out of the loop. Then just log that we couldn't allocate the structs required to continue.

Reply:
^

Reply:
I just made a change as per your suggestions. I ended up using a goto flag as using a break would exit the for loop. If I used a break in the first loop it will still enter the next loop.